src/data/finset/prod.lean

```diff
@@ -152,14 +152,21 @@ def diag := (s ×ˢ s).filter (λ a : α × α, a.fst = a.snd)
 for `a, b ∈ s`. -/
 def off_diag := (s ×ˢ s).filter (λ (a : α × α), a.fst ≠ a.snd)
 
-@[simp] lemma mem_diag (x : α × α) : x ∈ s.diag ↔ x.1 ∈ s ∧ x.1 = x.2 :=
+variables {s} {x : α × α}
+
+@[simp] lemma mem_diag : x ∈ s.diag ↔ x.1 ∈ s ∧ x.1 = x.2 :=
 by { simp only [diag, mem_filter, mem_product], split; intros h;
      simp only [h, and_true, eq_self_iff_true, and_self], rw ←h.2, exact h.1 }
 
-@[simp] lemma mem_off_diag (x : α × α) : x ∈ s.off_diag ↔ x.1 ∈ s ∧ x.2 ∈ s ∧ x.1 ≠ x.2 :=
+@[simp] lemma mem_off_diag : x ∈ s.off_diag ↔ x.1 ∈ s ∧ x.2 ∈ s ∧ x.1 ≠ x.2 :=
```

Comment:
This is an iff, hence should have as little explicit arguments as possible. I am changing it here because I was annoyed at overputting underscores and brackets in term mode.

Reply:
I disagree with this - explicit arguments are often *more* useful for `iff` statements, when you are rewriting and need to specify which instance you want to rewrite this. In particular, when this occurs (which is often for non-trivial statements), it requires many uses of underscores, while in term mode (which is used much less often than rewrites), only a few are needed (in this case just one).

Reply:
What you're saying goes against the standard. Look for `finset.mem_` lemmas. Out of the 30 lemmas that show up when looking for `finset.mem_` in the mathlib, there are only `finset.mem_fin`, `finset.mem_sum`, `finset.mem_map'` that take an explicit argument.

Reply:
> The point is that you _never_ rewrite by hand that precisely

This isn't true, many times - mostly for formalising real-life maths (or really, anything remotely nontrivial) - this sort of rewrite is precisely what's needed. In fact, the exact lemmas you mention have caused issues for me purely because they don't allow `rw` precisely enough to be useful, meaning that the resulting proof is cluttered with underscores which wouldn't be needed. Even *if* this is the standard, your justification of "overputting underscores" is explicitly false, more often than not this is required because you don't want to rewrite other occurrences.

Reply:
I think we will be stuck on this debate in the absence of more examples from both sides, so if you don't accept my justification of avoiding underscore I will justify this change as matching the standard for finset lemmas.

Reply:
I don't believe this justification is true either: there are plenty of finset lemmas - including `simp` ones, equalities and bi-implications - which have explicit data arguments. For some examples, consider `finset.disj_union_comm`, `finset.insert_eq`, `finset.insert_comm`, `finset.pair_eq_singleton`, `finset.inter_union_self`, or really any of the inter/union distrib lemmas

Reply:
And those all have `=` as their head, as opposed to `↔`.

Reply:
From my point of view, I would prefer this change unrelated to core functionality of this PR was not a blocker for it (as it clearly is currently).

Reply:
Given the discussion so far, I think it's clear that this PR interleaves a refactor with some new features. Would you be willing to split off the PR that only adds the `data/set/prod` features @YaelDillies ?